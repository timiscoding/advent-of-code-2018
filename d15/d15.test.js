const testData = require("./testData");
const { Game } = require("./Game");
const { Goblin, Elf } = require("./Player");

describe("Movement", () => {
  describe("When elf is equidistant to goblin", () => {
    it("it moves to first goblin in read order", () => {
      const g = new Game(testData.equidistantGoblins);
      g.players.goblins.push(new Goblin([1, 1], g.players, g.paths));
      const { nextPos, enemyPos } = g.players.elves[0].nextMove();
      expect(nextPos).toEqual([3, 1]);
      expect(enemyPos).toEqual([1, 1]);
    });

    it("same but with roles switched", () => {
      const g = new Game(testData.equidistantElves);
      g.players.elves.push(new Elf([1, 1], g.players, g.paths));
      const { nextPos, enemyPos } = g.players.goblins[0].nextMove();
      expect(nextPos).toEqual([3, 1]);
      expect(enemyPos).toEqual([1, 1]);
    });
  });

  describe("When elf has multiple equidistant paths to a goblin", () => {
    it("it moves to the square in read order", () => {
      const g = new Game(testData.equidistantMove);
      const { nextPos } = g.players.elves[0].nextMove();
      expect(nextPos).toEqual([3, 1]);
    });

    it("as above #2", () => {
      const g = new Game(testData.equidistantMove2);
      const { nextPos } = g.players.elves[0].nextMove();
      expect(nextPos).toEqual([3, 1]);
    });
  });

  it("When elf is right next to goblin, it doesn't move", () => {
    const g = new Game(testData.noMove);
    const { length } = g.players.elves[0].nextMove();
    expect(length).toBe(0);
  });

  describe("Multiple rounds", () => {
    it("When a player moves, its position updates", () => {
      const g = new Game(testData.playerMoves);
      let i = 3;
      let output = "\n";
      while (i--) {
        g.playRound();
        output += g.print({ silent: true }) + "\n";
      }
      expect(output).toMatchSnapshot();
    });

    it("as above #2", () => {
      const g = new Game(testData.playerMoves2);
      let i = 3;
      let output = "\n";
      while (i--) {
        g.playRound();
        output += g.print({ silent: true }) + "\n";
      }
      expect(output).toMatchSnapshot();
    });
  });
});

describe("Attack", () => {
  it("When player is adj to friends & enemies, it attacks enemy only", () => {
    const g = new Game(testData.attackEnemyOnly);
    const enemy = g.players.goblins[2].findAdjEnemy();
    expect(enemy).toBeInstanceOf(Elf);
  });

  it("When player is adj to friends, it doesn't attack", () => {
    const g = new Game(testData.attackNoone);
    const enemy = g.players.goblins[2].findAdjEnemy();
    expect(enemy).toBeNull();
  });

  it("When player is adj to noone, it doesn't attack", () => {
    const g = new Game(testData.attackNoone2);
    const enemy = g.players.goblins[0].findAdjEnemy();
    expect(enemy).toBeNull();
  });

  it("When player is adj to enemies, it attacks the weakest", () => {
    const g = new Game(testData.attackWeakest);
    g.players.elves[0].hp = 4;
    g.players.elves[1].hp = 3;
    g.players.elves[2].hp = 2;
    g.players.elves[3].hp = 4;
    const enemy = g.players.goblins[0].findAdjEnemy();
    expect(enemy).toBe(g.players.elves[2]);
  });

  it("When player is adj to enemies with same hp, it attacks the first in read order", () => {
    let g = new Game(testData.attackSameHP);
    let enemy = g.players.goblins[0].findAdjEnemy();
    expect(enemy).toBe(g.players.elves[0]);

    g = new Game(testData.attackSameHP2);
    enemy = g.players.goblins[0].findAdjEnemy();
    expect(enemy).toBe(g.players.elves[0]);

    g = new Game(testData.attackSameHP3);
    enemy = g.players.goblins[0].findAdjEnemy();
    expect(enemy).toBe(g.players.elves[0]);
  });

  it("When player attacks enemy, that enemy's hp reduces", () => {
    g = new Game(testData.attack);
    const elf = g.players.elves[0];
    const goblin = g.players.goblins[0];
    expect(elf.hp).toBe(200);
    expect(goblin.hp).toBe(200);
    g.playRound();
    expect(elf.hp).toBe(197);
    expect(goblin.hp).toBe(197);
  });

  describe("When player gets attacked", () => {
    it("and has 0 hp, that enemy should be removed from game", () => {
      g = new Game(testData.attack);
      const elf = g.players.elves[0];
      const goblin = g.players.goblins[0];
      elf.hp = 3;
      expect(elf.hp).toBe(3);
      expect(goblin.hp).toBe(200);
      g.playRound();
      expect(elf.hp).toBe(0);
      expect(goblin.hp).toBe(200);
      expect(g.players.elves).not.toContain(elf);
      expect(g.players.pos.has(g.paths.getVertex([elf.pos]))).toBe(false);
    });

    it("and has < 0 hp, it should be removed from game", () => {
      g = new Game(testData.attack);
      const elf = g.players.elves[0];
      const goblin = g.players.goblins[0];
      elf.hp = 1;
      expect(elf.hp).toBe(1);
      expect(goblin.hp).toBe(200);
      g.playRound();
      expect(elf.hp).toBe(-2);
      expect(goblin.hp).toBe(200);
      expect(g.players.elves).not.toContain(elf);
      expect(g.players.pos.has(g.paths.getVertex([elf.pos]))).toBe(false);
    });
  });
});
