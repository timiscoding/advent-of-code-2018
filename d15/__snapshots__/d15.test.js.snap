// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Attack Help the elves win! Sample 1 1`] = `
"#######
#..E..# E1(158)
#...E.# E0(14)
#.#.#.#
#...#.#
#.....#
#######
"
`;

exports[`Attack Help the elves win! Sample 2 1`] = `
"#######
#.E.E.# E0(200) E1(23)
#.#E..# E2(200)
#E.##E# E3(125) E4(200)
#.E.#.# E5(200)
#...#.#
#######
"
`;

exports[`Attack Help the elves win! Sample 3 1`] = `
"#######
#.E.#.# E0(8)
#.#E..# E1(86)
#..#..#
#...#.#
#.....#
#######
"
`;

exports[`Attack Help the elves win! Sample 4 1`] = `
"#######
#...E.# E0(14)
#.#..E# E1(152)
#.###.#
#.#.#.#
#...#.#
#######
"
`;

exports[`Attack Help the elves win! Sample 5 1`] = `
"#########
#.......#
#.E.#...# E0(38)
#..##...#
#...##..#
#...#...#
#.......#
#.......#
#########
"
`;

exports[`Attack Outcomes Sample 1 1`] = `
"#######
#G....# G0(200)
#.G...# G1(131)
#.#.#G# G2(59)
#...#.#
#....G# G3(200)
#######
"
`;

exports[`Attack Outcomes Sample 2 1`] = `
"#######
#...#E# E0(200)
#E#...# E2(197)
#.E##.# E5(185)
#E..#E# E4(200) E3(200)
#.....#
#######
"
`;

exports[`Attack Outcomes Sample 3 1`] = `
"#######
#.E.E.# E0(164) E2(197)
#.#E..# E4(200)
#E.##.# E3(98)
#.E.#.# E5(200)
#...#.#
#######
"
`;

exports[`Attack Outcomes Sample 4 1`] = `
"#######
#G.G#.# G2(200) G0(98)
#.#G..# G1(200)
#..#..#
#...#G# G3(95)
#...G.# G4(200)
#######
"
`;

exports[`Attack Outcomes Sample 5 1`] = `
"#######
#.....#
#.#G..# G2(200)
#.###.#
#.#.#.#
#G.G#G# G0(98) G1(38) G3(200)
#######
"
`;

exports[`Attack Outcomes Sample 6 1`] = `
"#########
#.G.....# G0(137)
#G.G#...# G3(200) G1(200)
#.G##...# G2(200)
#...##..#
#.G.#...# G4(200)
#.......#
#.......#
#########
"
`;

exports[`Movement Multiple rounds complex movement 1`] = `
"
#########
#.G...G.# G0(200) G2(200)
#...G...# G1(197)
#...E..G# E0(200) G4(200)
#.G.....# G3(200)
#.......#
#G..G..G# G5(200) G6(200) G7(200)
#.......#
#########

#########
#..G.G..# G0(200) G2(200)
#...G...# G1(194)
#.G.E.G.# G3(200) E0(197) G4(200)
#.......#
#G..G..G# G5(200) G6(200) G7(200)
#.......#
#.......#
#########

#########
#.......#
#..GGG..# G0(200) G1(191) G2(200)
#..GEG..# G3(200) E0(185) G4(200)
#G..G...# G5(200) G6(200)
#......G# G7(200)
#.......#
#.......#
#########

"
`;

exports[`Movement Multiple rounds simple movement 1`] = `
"
#######
#G..E.# G0(200) E0(200)
#.....#
#######

#######
#.GE..# G0(197) E0(200)
#.....#
#######

#######
#.GE..# G0(194) E0(197)
#.....#
#######

"
`;
